src/memtest.rs

```diff
@@ -0,0 +1,456 @@
+use {
+    crate::{prelude::*, TimeoutChecker},
+    rand::random,
+    std::{
+        error::Error,
+        fmt,
+        ptr::{read_volatile, write_bytes, write_volatile},
+    },
+};
+
+// TODO: Intend to convert this module to a standalone `no_std` crate
+// TODO: TimeoutChecker will be a trait instead
+
+#[derive(Debug)]
+pub enum MemtestOutcome {
+    Pass,
+    Fail(MemtestFailure),
+}
+
+#[derive(Debug)]
+pub enum MemtestFailure {
+    /// Failure due to the actual value read being different from the expected value
+    UnexpectedValue {
+        address: usize,
+        expected: usize,
+        actual: usize,
+    },
+    /// Failure due to the two memory locations being compared returning two different values
+    /// This is used by tests where memory is splitted in two and random values are written in pairs
+    MismatchedValues {
+        address1: usize,
+        value1: usize,
+        address2: usize,
+        value2: usize,
+    },
+}
+
+#[derive(Debug)]
+pub enum MemtestError {
+    Timeout,
+    Other(anyhow::Error),
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum MemtestType {
```

Comment:
Having the word `Type` can get a little screwy (it makes people think it might be a type alias). Consider calling this `MemtestKind` instead (just like `std::io::ErrorKind` did). I'm surprised Clippy didn't get mad at you for this, but get rid of the `Test` prefix on all your variants.